File: src/routes/reportRoutes.js
const express = require('express');
const router = express.Router();

// Sửa Lỗi 4: Đổi tên thư mục Middleware thành middleware (viết thường chữ m) để chạy đúng trên Linux
const AuthMiddleware = require('../middleware/authMiddleware');
const ReportController = require('../controllers/reportController');

router.post('/reports/generate', AuthMiddleware.rateLimiter, ReportController.generateReport);

// Sửa Lỗi 5: Thay thế hoàn toàn child_process.exec bằng fs.promises.readFile để chặn đứng lỗi Command Injection
router.get('/reports/lines', async (req, res) => {
  const fileName = req.query.file;
  
  if (!fileName) {
    return res.status(400).json({ error: "Filename parameter is required" });
  }

  try {
    const path = require('path');
    const fs = require('fs');
    const safeBaseDir = path.resolve(__dirname, '../data/reports');
    const targetPath = path.resolve(safeBaseDir, fileName);

    // Chặn Path Traversal cho API này
    if (!targetPath.startsWith(safeBaseDir)) {
      return res.status(403).json({ error: "Access denied." });
    }

    if (!fs.existsSync(targetPath)) {
      return res.status(404).json({ error: "File not found." });
    }

    const content = await fs.promises.readFile(targetPath, 'utf8');
    const linesCount = content.split('\n').length;
    res.json({ result: `${linesCount} ${fileName}` });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: src/routes/report_routes.py
from pathlib import Path
from fastapi import APIRouter, Query, HTTPException, Depends, Request
from src.middleware.auth_middleware import AuthMiddleware
from src.controllers.report_controller import ReportController

router = APIRouter()

@router.post("/reports/generate", dependencies=[Depends(AuthMiddleware.rate_limiter)])
def generate_report(request: Request):
    user = getattr(request.state, "user", None)
    return ReportController.generate_report(user)

@router.get("/reports/lines")
async def get_report_lines(file: str = Query(None)):
    if not file:
        raise HTTPException(status_code=400, detail="Filename parameter is required")
        
    try:
        current_dir = Path(__file__).resolve().parent
        safe_base_dir = (current_dir / ".." / "data" / "reports").resolve()
        target_path = (safe_base_dir / file).resolve()
        
        try:
            target_path.relative_to(safe_base_dir)
        except ValueError:
            raise HTTPException(status_code=403, detail="Access denied.")
            
        if not target_path.exists() or not target_path.is_file():
            raise HTTPException(status_code=404, detail="File not found.")
            
        content = target_path.read_text(encoding="utf-8")
        lines_count = len(content.split('\n'))
        return {"result": f"{lines_count} {file}"}
        
    except HTTPException:
        raise
    except Exception as error:
        raise HTTPException(status_code=500, detail=str(error))
